src/wallet/feebumper.cpp

```diff
@@ -210,6 +194,109 @@ Result CreateTransaction(const CWallet* wallet, const uint256& txid, const CCoin
+    if (result != Result::OK) {
+        return result;
+    }
+
+    // Fill in recipients(and preserve a change key)
+    std::vector<CRecipient> recipients;
+    for (const auto& output : wtx.tx->vout) {
+        if (!wallet->IsChange(output)) {
+            CRecipient recipient = {output.scriptPubKey, output.nValue, false};
+            recipients.push_back(recipient);
+        } else {
+            CTxDestination change_dest;
+            ExtractDestination(output.scriptPubKey, change_dest);
+            new_coin_control.destChange = change_dest;
+        }
+    }
+
+    // Old transaction stuff
+    // return old fee first
+    old_fee = wtx.GetDebit(ISMINE_SPENDABLE) - wtx.tx->GetValueOut();
+    int64_t txSize = GetVirtualTransactionSize(*(wtx.tx));
+    // Feerate of thing we are bumping
+    CFeeRate old_feerate(old_fee, txSize);
+
+    // The wallet uses a conservative WALLET_INCREMENTAL_RELAY_FEE value to
+    // future proof against changes to network wide policy for incremental relay
+    // fee that our node may not be aware of.
+    CFeeRate nodeIncrementalRelayFee = wallet->chain().relayIncrementalFee();
+    CFeeRate walletIncrementalRelayFee = CFeeRate(WALLET_INCREMENTAL_RELAY_FEE);
+    if (nodeIncrementalRelayFee > walletIncrementalRelayFee) {
+        walletIncrementalRelayFee = nodeIncrementalRelayFee;
+    }
+
+    // Compute required fee rate / total fee to hit
+    // We must be >= to this rate, and since tx will be as large or larger we should pass bip125 total fee check
+    CFeeRate min_new_feerate(old_feerate.GetFeePerK() + 1 + walletIncrementalRelayFee.GetFeePerK());
+
+    // Feerate calculated by GetMinimumFeeRate may not actually hit bip125 requirements
+    // so pick the larger of the two
+    CFeeRate block_feerate(GetMinimumFeeRate(*wallet, new_coin_control, /* feeCalc */ nullptr));
+    if (block_feerate.GetFeePerK() > min_new_feerate.GetFeePerK()) {
+        min_new_feerate = block_feerate;
+    }
+    // Finally, set required feerate.
+    new_coin_control.m_feerate = min_new_feerate;
+
+    // Fill in required inputs we are double-spending(all of them)
+    // N.B.: bip125 doesn't require all the inputs in the replaced transaction to be
+    // used in the replacement transaction, but it's very important for wallets to make
+    // sure that happens. If not, it would be possible to bump a transaction A twice to
+    // A2 and A3 where A2 and A3 don't conflict (or alternatively bump A to A2 and A2
+    // to A3 where A and A3 don't conflict). If both later get confirmed then the sender
+    // has accidentally double paid.
+    for (const auto& inputs : wtx.tx->vin) {
+        new_coin_control.Select(COutPoint(inputs.prevout));
+    }
+    new_coin_control.fAllowOtherInputs = true;
+
+    // We cannot source new unconfirmed inputs(bip125 rule 2)
+    new_coin_control.m_min_depth = 1;
+
+    CTransactionRef tx_new = MakeTransactionRef();
+    CReserveKey reservekey(wallet);
```

Comment:
I haven't been able to dig into this fully, but do you need to call `CReserveKey::KeepKey()` on this reserve key to remove it from the key pool after using it?

Reply:
good catch, fixed.

Reply:
https://github.com/bitcoin/bitcoin/pull/15557#discussion_r272314642

> good catch, fixed.

This was a pretty serious bug right? At least a privacy leak? Or am I mistaken?

Assuming this was serious, it seems like we should do something to prevent this happening in the future, like asserting in the `CReserveKey` destructor that either KeepKey or Reservekey was previously called.

Reply:
I think this would happen if you bumped the transaction, created a new change output, and then make another change output before the next block is seen. Otherwise the keypool-scanning we do would have taken care of it as soon as it's seen in a block.

A more conservative change that doesn't involve asserting would be to replace the `ReturnKey` call in the destructor to `KeepKey`, the idea being it's ok to burn a couple keys accidentally rather than repeat a script.

Reply:
The worst possible outcome would have been reusing a keypool address for change in multiple transactions. It's a marginal privacy leak but not disastrous.

The whole `CReserveKey` `GetReservedKey()`-`ReturnKey()`-`KeepKey()` cycle seems pretty brittle to me, and I think should be refactored. I haven't figured out exactly how yet, but I have a PR here: #15777 that documents how it currently works.